### PR TITLE
[SECURITY] Fix several cross-site scripting flaws

### PR DESCRIPTION
**Pull request by tyll**
Advisory: https://www.redteam-pentesting.de/advisories/rt-sa-2016-007

The advisory is not yet public, but it is scheduled to be released soon.

### CODE REVIEW

File: Classes/Controller/Form.php [hunk 1]
```diff
@@ -1109,6 +1109,9 @@ protected function init()
 
         $this->gp = $this->utilityFuncs->getMergedGP();
 
+        if (!$this->settings['uniqueFormID']) {
```
Comment (opi99) [inline Classes/Controller/Form.php:1112]: If I understand this correctly, if 'uniqueFormID' is set, we have the same issue.
So removing this if statement?

File: Classes/View/Form.php [hunk 2]
```diff
@@ -439,39 +439,39 @@ protected function fillDefaultMarkers()
         }
         $markers['###HIDDEN_FIELDS###'] = '
 			<input type="hidden" name="id" value="' . $GLOBALS['TSFE']->id . '" />
-			<input type="hidden" name="' . $name . '" value="1" />
+			<input type="hidden" name="' . htmlspecialchars($name) . '" value="1" />
```
Comment (opi99) [inline Classes/View/Form.php:442]: Why hsc this? FormValuesPrefix is a formhandler setting and isn't influenced by post/get vars?

File: Classes/Controller/Form.php [hunk 3]
```diff
@@ -1109,6 +1109,9 @@ protected function init()
...
+            $this->gp['randomID'] = preg_replace('/[^0-9a-z]/', '', preg_quote($this->gp['randomID']));
```
Comment (Phaeilo) [inline Classes/Controller/Form.php:1113]: `preg_quote` seems superfluous.

File: Classes/Controller/Form.php (hunk 1 above)
Comment (Phaeilo) [reply Classes/Controller/Form.php:1112]: Maybe move the `preg_replace` to line 1120 and sanitize `$randomID` unconditionally.

File: Classes/View/Form.php (hunk 2 above)
Comment (Phaeilo) [reply Classes/View/Form.php:442]: Maybe @reinhardfuehricht can explain this. In some other places `$name` isn't getting escaped currently.

**Comment by SvenJuergens** [comment]
@tyll @Phaeilo is this ready to merge?

File: Classes/Controller/Form.php (hunk 3 above)
Comment (gizmo21) [reply Classes/Controller/Form.php:1113]: I know it is deprecated, obsolete and unmaintained but would a backport fix of v.2.4.1 to the Formhandler v2.0.2 (last one available for T3 6.2ELTS) be to "preg_replace / quote" and "htmlspecialchars" the same lines in the corresponding files in the older version?

formhandler_2.0.2\Classes\Controller\Tx_Formhandler_Controller_Form.php
and
formhandler_2.0.2\Classes\View\Tx_Formhandler_View_Form.php

Comment (opi99) [reply Classes/Controller/Form.php:1113]: Formhandler 2.0.2 released to TER should have them inside, or take a look by AoE https://github.com/AOEpeople/formhandler

Comment (gizmo21) [reply Classes/Controller/Form.php:1113]: Yeah that's 2.0.2 but the old version with no fixes (like htmlspecialchars($name)) inside.

**Comment by tyll** [comment]
FYI: Unfortunately I do not have the time to take deeper looks into this issue.

File: Classes/Controller/Form.php (hunk 3 above)
Comment (opi99) [reply Classes/Controller/Form.php:1113]: Then check the forks ( https://github.com/KaffDaddy/formhandler )

Comment (gizmo21) [reply Classes/Controller/Form.php:1113]: Thanks opi99 for that one. That's at least exactly what I have done:
https://github.com/KaffDaddy/formhandler/commit/366e728b81d2a7c9c83bb29e8522a8eab736df33

So it seems that is best one can do on a 6.2ELTS release.

Also thanks tyll for discovering it 1year ago.

**Comment by opi99** [comment]
It's not complete but merging

**Pull request merged**